File: app/src/main/java/br/com/netflix/MainActivity.java
package br.com.netflix;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import br.com.netflix.model.Category;
import br.com.netflix.model.Movie;
import br.com.netflix.util.CategoryTask;
import br.com.netflix.util.ImageDownloaderTask;

public class MainActivity extends AppCompatActivity implements CategoryTask.CategoryLoader {

    private MainAdapter mainAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        _recycleViewInit();

        CategoryTask categoryTask = new CategoryTask(this);
        categoryTask.execute("https://tiagoaguiar.co/api/netflix/home");
        categoryTask.setCategoryLoader(this);
    }

    private void _recycleViewInit() {
        RecyclerView recyclerView = findViewById(R.id.recycler_view_main);
        List<Category> categories = new ArrayList<>();

        this.mainAdapter = new MainAdapter(categories);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
        recyclerView.setAdapter(this.mainAdapter);

    }

    @Override
    public void onResult(List<Category> categories) {
        this.mainAdapter.setCategories(categories);
        this.mainAdapter.notifyDataSetChanged();
    }

    private static class CategoryHolder extends RecyclerView.ViewHolder {

        private final TextView textViewTitle;
        private final RecyclerView recyclerViewMovies;

        public CategoryHolder(@NonNull View itemView) {
            super(itemView);
            textViewTitle = itemView.findViewById(R.id.text_view_title);
            recyclerViewMovies = itemView.findViewById(R.id.recycler_view_movie);
        }
    }

    private class MainAdapter extends RecyclerView.Adapter<CategoryHolder> {

        private List<Category> categories;

        public MainAdapter(List<Category> categories) {
            this.categories = categories;
        }

        public void setCategories(List<Category> categories) {
            this.categories.clear();
            this.categories.addAll(categories);
        }

        @NonNull
        @Override
        public CategoryHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            return new CategoryHolder(getLayoutInflater().inflate(R.layout.category_item, parent, false));
        }

        @Override
        public void onBindViewHolder(@NonNull CategoryHolder holder, int position) {
            Category category = categories.get(position);
            holder.textViewTitle.setText(category.getTitle());
            holder.recyclerViewMovies.setAdapter(new MovieAdapter(category.getMovies()));
            holder.recyclerViewMovies.setLayoutManager(new LinearLayoutManager(getBaseContext(), RecyclerView.HORIZONTAL, false));
        }

        @Override
        public int getItemCount() {
            return categories.size();
        }
    }


    private static class MovieHolder extends RecyclerView.ViewHolder {

        private final ImageView imageViewCover;
        private final ProgressBar progressBar;

        public MovieHolder(@NonNull View itemView, final OnItemClickListener onItemClickListener) {
            super(itemView);
            imageViewCover = itemView.findViewById(R.id.image_view_cover);
            progressBar = itemView.findViewById(R.id.movie_item_progress);

            itemView.setOnClickListener((view) -> onItemClickListener.onClick(getAdapterPosition()));
        }
    }

    private class MovieAdapter extends RecyclerView.Adapter<MovieHolder> implements OnItemClickListener {

        private final List<Movie> movies;

        public MovieAdapter(List<Movie> movies) {
            this.movies = movies;
        }

        @Override
        public void onClick(int position) {
            Intent intent = new Intent(MainActivity.this, MovieInfoActivity.class);
            int id = movies.get(position).getId();

            if (id > 0 && id <= 4) {
                intent.putExtra("id", id);
                startActivity(intent);
            }
        }

        @NonNull
        @Override
        public MovieHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = getLayoutInflater().inflate(R.layout.movie_item, parent, false);
            return new MovieHolder(view, this);
        }

        @Override
        public void onBindViewHolder(@NonNull MovieHolder holder, int position) {
            Movie movie = movies.get(position);
            new ImageDownloaderTask(holder.progressBar, holder.imageViewCover, null, false).execute(movie.getCoverUrl());
        }

        @Override
        public int getItemCount() {
            return movies.size();
        }


    }

    interface OnItemClickListener {
        void onClick(int position);
    }
}

File: app/src/main/java/br/com/netflix/kotlin/MainActivity.kt
package br.com.netflix.kotlin

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import br.com.netflix.R
import br.com.netflix.model.Categories
import br.com.netflix.model.Category
import br.com.netflix.model.Movie
import com.bumptech.glide.Glide
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.category_item.view.*
import kotlinx.android.synthetic.main.movie_item.view.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class MainActivity : AppCompatActivity() {
    private lateinit var mainAdapter: MainAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val categories: ArrayList<Category> = arrayListOf()
        mainAdapter = MainAdapter(categories)
        recycler_view_main.adapter = mainAdapter
        recycler_view_main.layoutManager = LinearLayoutManager(this)

//        val catgoryTask = CategoryTask(this)
//        catgoryTask.setCategoryLoader {
//            mainAdapter.categories.clear()
//            mainAdapter.categories.addAll(it)
//            mainAdapter.notifyDataSetChanged()
//        }
//        catgoryTask.execute("https://tiagoaguiar.co/api/netflix/home")

        retrofit().create(NetflixAPI::class.java)
            .getCategories()
            .enqueue(object : Callback<Categories> {
                override fun onResponse(call: Call<Categories>, response: Response<Categories>) {
                    if (response.isSuccessful) {
                        response.body()?.let {
                            mainAdapter.categories.clear()
                            mainAdapter.categories.addAll(it.categories)
                            mainAdapter.notifyDataSetChanged()
                        }
                    }
                }

                override fun onFailure(call: Call<Categories>, t: Throwable) {
                    Toast.makeText(this@MainActivity, t.message, Toast.LENGTH_SHORT)
                }

            })
    }

    private inner class MainAdapter(val categories: MutableList<Category>) :
        RecyclerView.Adapter<CategoryHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryHolder =
            CategoryHolder(layoutInflater.inflate(R.layout.category_item, parent, false))


        override fun onBindViewHolder(holder: CategoryHolder, position: Int) {
            holder.bind(categories[position])
        }

        override fun getItemCount(): Int = categories.size


    }

    private inner class CategoryHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(category: Category) = with(itemView) {
            text_view_title.text = category.title
            recycler_view_movie.adapter = MovieAdapter(category.movies) {
                if (it.id > 3) {
                    Toast.makeText(this@MainActivity, "Filme não cadastrado!", Toast.LENGTH_LONG)
                        .show()
                } else {
                    val intent = Intent(this@MainActivity, MovieInfoActivity::class.java)
                    intent.putExtra("id", it.id)
                    startActivity(intent)
                }
            }
            recycler_view_movie.layoutManager =
                LinearLayoutManager(this@MainActivity, RecyclerView.HORIZONTAL, false)
        }

    }


    private inner class MovieAdapter(
        val movies: List<Movie>,
        private val listener: ((Movie) -> Unit)?
    ) :
        RecyclerView.Adapter<MovieHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieHolder =
            MovieHolder(
                layoutInflater.inflate(R.layout.movie_item, parent, false),
                listener
            )


        override fun onBindViewHolder(holder: MovieHolder, position: Int) =
            holder.bind(movies[position])


        override fun getItemCount(): Int = movies.size


    }

    private class MovieHolder(itemView: View, val onClick: ((Movie) -> Unit)?) :
        RecyclerView.ViewHolder(itemView) {
        fun bind(movie: Movie) = with(itemView) {
//            ImageDownloaderTask(
//                movie_item_progress,
//                image_view_cover,
//                null,
//                false
//            ).execute(movie.coverUrl)
            Glide.with(context)
                .load(movie.coverUrl)
                .placeholder(R.drawable.placeholder_bg)
                .into(image_view_cover)

            image_view_cover.setOnClickListener {
                onClick?.invoke(movie)
            }
        }
    }
}